README.md
# Holly
> 🌿 o!rdr command line tool

Holly is a simple command line tool which allows you to upload osu! replays
to [o!rdr](https://ordr.issou.best/) quickly and easily.  

https://user-images.githubusercontent.com/38820196/149671948-97f59a68-35f1-4960-a71f-84fdbd4f03a5.mp4

# Install
### Releases
Check out the [latest release](https://github.com/TorchedSammy/Holly/releases/latest)
and choose whichever fits your platform.

### Compile
```
git clone https://github.com/TorchedSammy/Holly
cd Holly
cargo build
```

# Usage
Using Holly is very simple: run it, with a provided path to a replay file.  

# License
MIT



CHANGELOG.md
# Changelog
This project adheres to [Semantic Versioning](https://semver.org).

## [0.2.0] - 2022-01-16
- Rewrite in rust

## [0.1.0] - 2021-11-09

Initial release

[0.2.0]: https://github.com/TorchedSammy/Holly/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/TorchedSammy/Holly/releases/tag/v0.1.0


main.go
package main

import (
	"bytes"
	"fmt"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"

	flag "github.com/spf13/pflag"
)

func main() {
	// Flag for skin of replay (and default to blueberry, the best skin)
	var skinFlag = flag.StringP("skin", "s", "blueberry_v1_7_0", "Skin to use")
	flag.Parse()

	replayFile := flag.Arg(0)
	if replayFile == "" {
		fmt.Println("Missing file path to the replay..")
		os.Exit(1)
	}

	params := map[string]string{
		"skin": *skinFlag,
		"username": "Holly",
		"resolution": "1280x720",
	}

	req, err := UploadReplay(replayFile, params)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	// Sends our request
	client := &http.Client{}
	resp, err := client.Do(req)

	if err != nil {
		fmt.Println(err)
	} else {
		var bodyContent = make([]byte, req.ContentLength)
		fmt.Println(resp.StatusCode)
		resp.Body.Read(bodyContent)
		resp.Body.Close()
		fmt.Println(string(bodyContent))
	}
}

func UploadReplay(path string, params map[string]string) (*http.Request, error) {
	f, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	body := &bytes.Buffer{}

	w := multipart.NewWriter(body)
	// Add our file field
	part, err := w.CreateFormFile("replayFile", filepath.Base(path))
	if err != nil {
		return nil, err
	}
	part.Write(f) // Write replay content

	// Do the other params
	for key, val := range params {
		_ = w.WriteField(key, val)
	}

	err = w.Close()
	if err != nil {
		return nil, err
	}

	req, _ := http.NewRequest("POST", "https://ordr-api.issou.best/renders", body)
	req.Header.Add("Content-Type", w.FormDataContentType()) // Set content-type to be multipart
	return req, nil
}



src/main.rs
mod holly;

use indicatif::ProgressBar;
use reqwest::blocking::multipart;
use rust_socketio::{Client, ClientBuilder, Payload};
use std::io::Read;
use std::{env, time};

fn main() {
	let args: Vec<String> = env::args().collect();
	let spinner = ProgressBar::new_spinner(); // thanks indicatif this makes sense
	spinner.enable_steady_tick(80);

	let form = multipart::Form::new()
		.text("skin", "blueberry_v1_7_0")
		.text("resolution", "1280x720")
		.text("username", "Holly")
		.text("showDanserLogo", "false")
		.file("replayFile", &args[1])
		.unwrap();

	spinner.set_message("Uploading replay...");
	let client = reqwest::blocking::Client::new();
	let mut res = client
		.post("https://ordr-api.issou.best/renders")
		.multipart(form)
		.send()
		.unwrap();
	spinner.set_message("Replay uploaded, waiting for response");

	// check if status code is 429
	if res.status() == reqwest::StatusCode::TOO_MANY_REQUESTS {
		// get X-RateLimit-Reset header
		let now = time::Instant::now();
		let stamp = res
			.headers()
			.get("X-RateLimit-Reset")
			.unwrap()
			.to_str()
			.unwrap()
			.parse::<u64>()
			.unwrap();
		spinner.finish_and_clear();
		println!("Ratelimited, try again later...");
	}

	let mut body = String::new();
	res.read_to_string(&mut body).unwrap();

	let r: holly::SentRender = serde_json::from_str(&body).unwrap();

	// on 429 it doesn't send json so we check if response code isnt 201 and
	// print error message from json
	if res.status() != reqwest::StatusCode::CREATED {
		println!("Error: {} ({})", r.message, r.errorCode);
		return;
	}
	spinner.set_message("Uploaded successfully, waiting for render to start");
	let render_id = r.renderID;

	let done_callback = {
		let spinner = spinner.clone();
		move |payload: Payload, socket: Client| {
			let render_id = render_id;
			let data = match payload {
				Payload::String(s) => s,
				_ => "".to_string(),
			};

			let p: holly::RenderDone = serde_json::from_str(&data).unwrap();
			if p.renderID == render_id {
				spinner.set_message(p.videoUrl);
				spinner.finish();
				socket.disconnect();
				std::process::exit(0);
			}
		}
	};

	let progress_callback = {
		let spinner = spinner.clone();
		move |payload: Payload, _: Client| {
			let data = match payload {
				Payload::String(s) => s,
				_ => "".to_string(),
			};
			let p: holly::RenderProgress = serde_json::from_str(&data).unwrap();
			if p.renderID == render_id {
				spinner.set_message(p.progress);
			}
		}
	};

	let failed_callback = {
		let spinner = spinner.clone();
		move |payload: Payload, socket: Client| {
			let data = match payload {
				Payload::String(s) => s,
				_ => "".to_string(),
			};

			let p: holly::RenderFailed = serde_json::from_str(&data).unwrap();
			if p.renderID == render_id {
				spinner.set_message(format!(
					"Render failed. Error: {} ({})",
					p.errorMessage, p.errorCode
				));
				spinner.finish();
				socket.disconnect();
				std::process::exit(0);
			}
		}
	};
	ClientBuilder::new("https://ordr-ws.issou.best")
		.on("render_progress_json", progress_callback)
		.on("render_done_json", done_callback)
		.on("render_failed_json", failed_callback)
		.connect()
		.expect("Connection failed");

    loop {}
}


src/index.js
const fs = require('fs');
const path = require('path');
const io = require('socket.io-client');
const fetch = require('node-fetch-commonjs');
const owy = require('owy');
const FormData = require('form-data');
const form = new FormData();
const spinner = new owy.Spinner();

const filePath = process.argv[2];
const fileStream = fs.createReadStream(filePath);

form.append('replayFile', fileStream, path.basename(filePath));
form.append('skin', 'blueberry_v1_7_0')
form.append('username', 'Holly')
form.append('resolution', '1280x720')

const options = {
    method: 'POST',
    body: form,
    headers: {
    	'Content-Type': `multipart/form-data; boundary=${form._boundary}`
    }
}

spinner.start('Uploading replay...');

function timeRelative(time) {
	const now = Date.now();
	const diff = time - now;
	const seconds = Math.floor(diff / 1000);
	const minutes = Math.floor(seconds / 60);
	
	if (minutes > 0) {
		return `in ${minutes} minutes, ${seconds % 60} seconds`;
	} else if (seconds > 0) {
		return `in ${seconds} seconds`;
	} else {
		return 'in less than a second';
	}
}

const res = fetch('https://ordr-api.issou.best/renders', options)
.then(async res => {
	if(res.status === 429) {
		spinner.fail(`Replay has been uploaded recently. Please wait ${timeRelative(Number(res.headers.get("X-RateLimit-Reset")) * 1000)} before trying again.`);
		process.exit(1);
	}

	const resp = await res.json();
	if(res.status !== 201) {
		spinner.fail(`Error: ${resp.message} (${res.status})`);
		process.exit(1);
	}
	spinner.success('Replay uploaded! Render will start in a few seconds.')

	const socket = io('https://ordr-ws.issou.best');
	spinner.start('Waiting for ordr...');
	socket.on('render_progress_json', renderInfo => {
		if(renderInfo.renderID === resp.renderID) {
			if(renderInfo.progress.includes('Finalizing')) {
				spinner.text = 'Video is being uploaded...';
				return
			}
			spinner.text = `${renderInfo.progress}`;
		}
	});

	socket.on('render_done_json', renderInfo => {
		if(renderInfo.renderID === resp.renderID) {
			socket.disconnect();
			spinner.success(renderInfo.videoUrl)
		}
	});

	socket.on('render_failed_json', renderInfo => {
		if(renderInfo.renderID === resp.renderID) {
			socket.disconnect();
			spinner.fail(`Error: ${renderInfo.errorMessage} (${renderInfo.errorCode})`);
		}
	});
});
